spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/CredentialsPropertiesSetter.java

```diff
@@ -0,0 +1,91 @@
+import com.google.cloud.spring.core.Credentials;
+import com.google.cloud.sql.CredentialFactory;
+import java.io.File;
+import java.io.IOException;
+import org.apache.commons.logging.Log;
+import org.springframework.core.io.Resource;
+
+/**
+ * Sets credentials to be used by GCP socket factory.
+ */
+final class CredentialsPropertiesSetter {
+
+  CredentialsPropertiesSetter() {
+  }
+
+  /**
+   * Set credentials to be used by the Google Cloud SQL socket factory.
+   *
+   * <p>The only way to pass a {@link CredentialFactory} to the socket factory is by passing a
+   * class name through a system property. The socket factory creates an instance of {@link
+   * CredentialFactory} using reflection without any arguments. Because of that, the credential
+   * location needs to be stored somewhere where the class can read it without any context. It could
+   * be possible to pass in a Spring context to {@link SqlCredentialFactory}, but this is a tricky
+   * solution that needs some thinking about.
+   *
+   * <p>If user didn't specify credentials, the socket factory already does the right thing by
+   * using the application default credentials by default. So we don't need to do anything.
+   */
+  static void setCredentials(GcpCloudSqlProperties sqlProperties, GcpProperties gcpProperties,
+      Log logger) {
+    Credentials credentials = null;
+
+    if (sqlProperties.getCredentials().hasKey()) {
+      // First tries the SQL configuration credential.
+      credentials = sqlProperties.getCredentials();
+    } else {
+      // Then, the global credential.
+      credentials = gcpProperties.getCredentials();
+    }
+
+    if (credentials.getEncodedKey() != null) {
+      setCredentialsEncodedKeyProperty(credentials.getEncodedKey());
+    } else if (credentials.getLocation() != null) {
+      setCredentialsFileProperty(credentials.getLocation(), logger);
+    }
+    // Else do nothing, let sockets factory use application default credentials.
+  }
+
+  private static void setCredentialsEncodedKeyProperty(String encodedKey) {
+    System.setProperty(SqlCredentialFactory.CREDENTIAL_ENCODED_KEY_PROPERTY_NAME,
+        encodedKey);
+
+    System.setProperty(CredentialFactory.CREDENTIAL_FACTORY_PROPERTY,
+        SqlCredentialFactory.class.getName());
+  }
+
+  private static void setCredentialsFileProperty(Resource credentialsLocation, Log logger) {
+    try {
+      // A resource might not be in the filesystem, but the Cloud SQL credential must.
+      File credentialsLocationFile = credentialsLocation.getFile();
+
+      System.setProperty(SqlCredentialFactory.CREDENTIAL_LOCATION_PROPERTY_NAME,
+          credentialsLocationFile.getAbsolutePath());
+
+      // If there are specified credentials, tell sockets factory to use them.
+      System.setProperty(CredentialFactory.CREDENTIAL_FACTORY_PROPERTY,
+          SqlCredentialFactory.class.getName());
+    } catch (IOException ioe) {
+      logger.info("Error reading Cloud SQL credentials file.", ioe);
```

Comment:
I think this failure should propagate back to the user. If the user provided custom credentials and they did not work, the application is not really expected to work. Unless falling back to default credentials is acceptable (it might be in some cases? but it's hard to assume).